### PR TITLE
trie: allow trie put/del ops to skip transforming key for snapsync

### PR DESCRIPTION
**Pull request by g11tech**
Snap sync requires building the MPT from the peer tree where key is already hashed. This PR allows trie put/del opts to skip key transform even if trie is set with `useKeyHashing`. This allows to correctly build the MPT and correctly map addresse lookups to the trie

TODO:
-  [x] add some account lookups in the test to validate trie is build correctly
-  [x] add unit cases for trie

### CODE REVIEW

**Comment by g11tech** [comment]
coverage for snap related fetchers will increase as we do followup PRs for integration

**Review by scorbajio — COMMENTED** [review]
This is looking great! One of the tests fail at the end, seemingly due to a missing account:
```
 FAIL  test/sim/snapsync.spec.ts > simple mainnet test run > should match entire state
TypeError: Cannot read properties of undefined (reading 'balance')
 ❯ test/sim/snapsync.spec.ts:201:17
    199|       console.log(addressString)
    200|       assert.equal(
    201|         account.balance,
       |                 ^
    202|         BigInt(customGenesisState[addressString][0]),
    203|         `${addressString} balance should match`

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/1]⎯

 Test Files  1 failed (1)
      Tests  1 failed | 3 passed | 1 skipped (5)
   Start at  13:59:28
   Duration  63.49s (transform 1.97s, setup 0ms, collect 43.23s, tests 19.60s, environment 0ms, prepare 213ms)
```
But since the roots match, I suspect it's not an issue with the fetchers. Does the test also fail for you when you run the sim test suit locally?

**Comment by g11tech** [comment]
> But since the roots match, I suspect it's not an issue with the fetchers. Does the test also fail for you when you run the sim test suit locally?

hey no, all tests pass for me... let me do some variations of the cmd, might be an edge case somewhere

File: packages/trie/src/trie.ts
Comment (ScottyPoi) [inline packages/trie/src/trie.ts:1]: So we're setting up a Trie where the key used for `put` and `del` is different than the key used for `get`?

File: packages/client/src/sync/fetcher/accountfetcher.ts [hunk 1]
```diff
@@ -376,7 +384,8 @@ export class AccountFetcher extends Fetcher<JobTask, AccountData[], AccountData>
     const storageFetchRequests = new Set()
     const byteCodeFetchRequests = new Set<Uint8Array>()
     for (const account of result) {
-      await this.accountTrie.put(account.hash, accountBodyToRLP(account.body))
+      // what we have is hashed account and not its pre-image, so we skipKeyTransform
+      await this.accountTrie.put(account.hash, accountBodyToRLP(account.body), true)
```
[the remark this answers is not in the document]
Comment (g11tech) [reply packages/client/src/sync/fetcher/accountfetcher.ts:388]: i think you got the keyvalue pair mixed up, first is key second is value

File: packages/trie/src/trie.ts
Comment (g11tech) [reply packages/trie/src/trie.ts:1]: correct because in snap sync we already get hashed key in the AccountData for e.g. (account.hash for e.g.)

**Comment by g11tech** [comment]
> This is looking great! One of the tests fail at the end, seemingly due to a missing account:
> 
> ```
>  FAIL  test/sim/snapsync.spec.ts > simple mainnet test run > should match entire state
> TypeError: Cannot read properties of undefined (reading 'balance')
>  ❯ test/sim/snapsync.spec.ts:201:17
>     199|       console.log(addressString)
>     200|       assert.equal(
>     201|         account.balance,
>        |                 ^
>     202|         BigInt(customGenesisState[addressString][0]),
>     203|         `${addressString} balance should match`
> 
> ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/1]⎯
> 
>  Test Files  1 failed (1)
>       Tests  1 failed | 3 passed | 1 skipped (5)
>    Start at  13:59:28
>    Duration  63.49s (transform 1.97s, setup 0ms, collect 43.23s, tests 19.60s, environment 0ms, prepare 213ms)
> ```
> 
> But since the roots match, I suspect it's not an issue with the fetchers. Does the test also fail for you when you run the sim test suit locally?

i fixed the edge case, thanks for spotting it :slightly_smiling_face:

File: packages/client/src/sync/fetcher/accountfetcher.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (scorbajio) [reply packages/client/src/sync/fetcher/accountfetcher.ts:388]: @ScottyPoi the second one should be `accountTrie.get(account.hash) = accountBodyToRLP(account.body)` where `account.hash` is the `keccak256` hash of the account address that is derived from the public key of the account's corresponding private key. The account key is **not** derived from the account body.

**Review by scorbajio — APPROVED** [review]
ACK

File: packages/trie/src/trie.ts
Comment (scorbajio) [reply packages/trie/src/trie.ts:1]: We are putting by post-image (and skipping the hashing from pre to post-image) and reading by pre-image, which would be the account address.

File: packages/client/src/sync/fetcher/accountfetcher.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ScottyPoi) [reply packages/client/src/sync/fetcher/accountfetcher.ts:388]: sweet, that makes perfect sense.  I was worried that we were mixing up the account address with the DB Key, that's all.

  LGTM, :+1:
